File: android/app/src/main/java/com/monivitals/source/DataSource.kt
//
//  DataSource.kt
//  MoniVitals (Android)
//
//  The single interface that the UI talks to. The replay source (real recorded biosignal)
//  and the synthetic mock source both implement it, so every screen works without hardware.
//  Packets delivered to consumers carry EXTENDED device timestamps (see Timebase).
//
//  Mirrors ios/MoniVitals/Source/DataSource.swift. Event delivery uses SharedFlow (the
//  Combine PassthroughSubject equivalent); handlers push high-rate samples into
//  non-reactive ring buffers rather than reactive state.
//

package com.monivitals.source

import com.monivitals.ble.BiozPacket
import com.monivitals.ble.EcgPacket
import com.monivitals.ble.InfoMessage
import com.monivitals.ble.MetricsPacket
import com.monivitals.ble.PpgPacket
import com.monivitals.ble.StatusMessage
import com.monivitals.ble.StreamKind
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow

/** Connection lifecycle, mirroring the web `ConnectionState` union. */
enum class ConnectionState {
    DISCONNECTED,
    UNSUPPORTED,
    SCANNING,
    CONNECTING,
    CONNECTED,
    RECONNECTING,
}

/** Optional detail attached to a connection-state change. */
data class ConnectionDetail(val error: String? = null)

private fun <T> streamFlow(): MutableSharedFlow<T> =
    MutableSharedFlow(extraBufferCapacity = 512, onBufferOverflow = BufferOverflow.DROP_OLDEST)

/**
 * SharedFlows a `DataSource` publishes on. The ViewModel collects these. Emission uses
 * `tryEmit` from any thread (BLE binder thread or timer thread), matching the buffered,
 * non-suspending semantics of a Combine PassthroughSubject.
 */
class DataSourceEvents {
    val ecg = streamFlow<EcgPacket>()
    val bioz = streamFlow<BiozPacket>()
    val ppg = streamFlow<PpgPacket>()

    /** Firmware-computed metrics (optional; the app computes its own in parallel). */
    val metrics = streamFlow<MetricsPacket>()
    val status = streamFlow<StatusMessage>()
    val info = streamFlow<InfoMessage>()
    val battery = streamFlow<Int>()

    /** (stream, newly dropped packet count) from seq gaps. */
    val dropped = streamFlow<Pair<StreamKind, Int>>()
    val connection = streamFlow<Pair<ConnectionState, ConnectionDetail?>>()
}

/** The data-source abstraction implemented by `ReplaySource` (real recorded biosignal) and
 *  `MockSource` (synthetic), so the app runs standalone with no hardware. */
interface DataSource {
    val isMock: Boolean
    val state: ConnectionState

    /** offset = hostNowUs − deviceTus, or null before SYNC_CLOCK. */
    val clockOffsetUs: Double?
    val deviceName: String?

    /** The event hub for this source. */
    val events: DataSourceEvents

    fun connect()
    fun disconnect()
    fun start(streamMask: Int)
    fun stop()
    fun setRate(stream: StreamKind, hz: Int)
    fun syncClock()
    fun getInfo()
}

/** Shared base providing the event hub + connection-state plumbing. */
abstract class BaseDataSource : DataSource {
    override val events = DataSourceEvents()

    private var stateValue: ConnectionState = ConnectionState.DISCONNECTED

    override val isMock: Boolean get() = false
    override val clockOffsetUs: Double? get() = null
    override val deviceName: String? get() = null

    override val state: ConnectionState get() = stateValue

    protected fun setState(state: ConnectionState, detail: ConnectionDetail? = null) {
        stateValue = state
        events.connection.tryEmit(state to detail)
    }

    // Subclasses override these.
    override fun connect() {}
    override fun disconnect() {}
    override fun start(streamMask: Int) {}
    override fun stop() {}
    override fun setRate(stream: StreamKind, hz: Int) {}
    override fun syncClock() {}
    override fun getInfo() {}
}

File: ios/MoniVitals/Source/DataSource.swift
//
//  DataSource.swift
//  MoniVitals
//
//  The single interface that the UI talks to. The signal sources (ReplaySource, which
//  plays the bundled real recording, and MockSource, a synthetic generator) implement it,
//  so every screen works without hardware. Packets delivered to consumers carry EXTENDED
//  device timestamps (see Timebase.swift).
//
//  Mirrors web/src/source/DataSource.ts. Event delivery uses Combine subjects so the
//  AppModel can subscribe with backpressure-free sinks (the high-rate packet handlers
//  push into non-reactive ring buffers, not @Published state).
//

import Combine
import Foundation

/// Connection lifecycle, mirroring the web `ConnectionState` union.
enum ConnectionState: String {
    case disconnected
    case unsupported
    case scanning
    case connecting
    case connected
    case reconnecting
}

/// Optional detail attached to a connection-state change.
struct ConnectionDetail {
    var error: String?
}

/// Combine subjects a `DataSource` publishes on. The AppModel subscribes to these.
/// Subjects are `Void`-failing PassthroughSubjects (events only, no completion).
final class DataSourceEvents {
    let ecg = PassthroughSubject<EcgPacket, Never>()
    let bioz = PassthroughSubject<BiozPacket, Never>()
    let ppg = PassthroughSubject<PpgPacket, Never>()
    /// Source-provided reference metrics (optional; the app computes its own in parallel).
    let metrics = PassthroughSubject<MetricsPacket, Never>()
    let status = PassthroughSubject<StatusMessage, Never>()
    let info = PassthroughSubject<InfoMessage, Never>()
    let battery = PassthroughSubject<Int, Never>()
    /// (stream, newly dropped packet count) from seq gaps.
    let dropped = PassthroughSubject<(StreamKind, Int), Never>()
    let connection = PassthroughSubject<(ConnectionState, ConnectionDetail?), Never>()
}

/// The data-source abstraction implemented by `ReplaySource` (bundled real recording) and
/// `MockSource` (standalone synthetic source).
protocol DataSource: AnyObject {
    var isMock: Bool { get }
    var state: ConnectionState { get }
    /// offset = hostNowUs − deviceTus, or nil before SYNC_CLOCK.
    var clockOffsetUs: Double? { get }
    var deviceName: String? { get }

    /// The Combine event hub for this source.
    var events: DataSourceEvents { get }

    func connect()
    func disconnect()
    func start(streamMask: UInt8)
    func stop()
    func setRate(_ stream: StreamKind, hz: Int)
    func syncClock()
    func getInfo()
}

/// Shared base providing the event hub + connection-state plumbing.
class BaseDataSource: DataSource {
    let events = DataSourceEvents()
    private(set) var stateValue: ConnectionState = .disconnected

    var isMock: Bool { false }
    var clockOffsetUs: Double? { nil }
    var deviceName: String? { nil }

    var state: ConnectionState { stateValue }

    func setState(_ state: ConnectionState, detail: ConnectionDetail? = nil) {
        stateValue = state
        events.connection.send((state, detail))
    }

    // Subclasses override these.
    func connect() {}
    func disconnect() {}
    func start(streamMask: UInt8) {}
    func stop() {}
    func setRate(_ stream: StreamKind, hz: Int) {}
    func syncClock() {}
    func getInfo() {}
}
